packages/dev/core/src/Rendering/fluidRenderer/fluidRenderingObjectParticleSystem.ts

```diff
@@ -0,0 +1,137 @@
+import type { Scene } from "core/scene";
+import type { Nullable } from "core/types";
+
+import { FluidRenderingObject } from "./fluidRenderingObject";
+
+/**
+ * Defines a rendering object based on a particle system
+ */
+export class FluidRenderingObjectParticleSystem extends FluidRenderingObject {
+    private _particleSystem: IParticleSystem;
+    private _originalRender: () => number;
+    private _renderDiffuse: () => number;
+    private _blendMode: number;
+    private _onBeforeDrawParticleObserver: Nullable<Observer<Nullable<Effect>>>;
+    private _updateInAnimate: boolean;
+
+    /** Gets the particle system */
+    public get particleSystem() {
+        return this._particleSystem;
+    }
+
+    /**
+     * Gets the name of the class
+     */
+    public getClassName(): string {
+        return "FluidRenderingObjectParticleSystem";
+    }
+
+    private _useTrueRenderingForDiffuseTexture = true;
+
+    /**
+     * Gets or sets a boolean indicating that the diffuse texture should be generated based on the regular rendering of the particle system (default: true).
+     * Sometimes, generating the diffuse texture this way may be sub-optimal. In that case, you can disable this property, in which case the particle system will be
+     * rendered using a ALPHA_COMBINE mode instead of the one used by the particle system.
+     */
+    public get useTrueRenderingForDiffuseTexture() {
+        return this._useTrueRenderingForDiffuseTexture;
+    }
+
+    public set useTrueRenderingForDiffuseTexture(use: boolean) {
+        if (this._useTrueRenderingForDiffuseTexture === use) {
+            return;
+        }
+
+        this._useTrueRenderingForDiffuseTexture = use;
+
+        if (use) {
+            this._particleSystem.blendMode = this._blendMode;
+            this._particleSystem.onBeforeDrawParticlesObservable.remove(this._onBeforeDrawParticleObserver);
+            this._onBeforeDrawParticleObserver = null;
+        } else {
+            this._particleSystem.blendMode = -1;
+            this._onBeforeDrawParticleObserver = this._particleSystem.onBeforeDrawParticlesObservable.add(() => {
+                this._engine.setAlphaMode(Constants.ALPHA_COMBINE);
+            });
+        }
+    }
+
+    /**
+     * Gets the vertex buffers
+     */
+    public get vertexBuffers(): { [key: string]: VertexBuffer } {
+        return this._particleSystem.vertexBuffers as { [key: string]: VertexBuffer };
+    }
+
+    /**
+     * Gets the index buffer (or null if the object is using instancing)
+     */
+    public get indexBuffer(): Nullable<DataBuffer> {
+        return this._particleSystem.indexBuffer;
+    }
+
+    /**
+     * Creates a new instance of the class
+     * @param scene The scene the particle system is part of
+     * @param ps The particle system
+     */
+    constructor(scene: Scene, ps: IParticleSystem) {
+        super(scene);
+
+        this._particleSystem = ps;
+
+        this._originalRender = ps.render.bind(ps);

```

Comment:
why do we keep both _originalRender and _renderDiffuse ? also why not calling this._particleSystem.rennder() ? I am not sure the indirections are necessary here ?

Reply:
I'm not calling this._particleSystem.render() because the function is "shut down" by `this._particleSystem.render = () => 0` => I don't want the render function to be called as part of the normal render flow (particleSystem.render() is called by  `RenderingGroup`).